core/src/tools/testfind_fd.cc

```diff
@@ -0,0 +1,285 @@
+
+   You should have received a copy of the GNU Affero General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+   02110-1301, USA.
+*/
+
+#include "dird/fd_sendfileset.h"
+#include "filed/filed_utils.h"
+#include "filed/filed.h"
+#include "filed/filed_jcr_impl.h"
+#include "filed/filed_globals.h"
+#include "filed/dir_cmd.h"
+#include "filed/fileset.h"
+#include "findlib/attribs.h"
+#include "lib/parse_conf.h"
+#include "tools/dummysockets.h"
+#include "tools/testfind_fd.h"
+
+using namespace filedaemon;
+
+static struct TestfindStats {
+  int num_files = 0;
+  int max_file_len = 0;
+  int max_path_len = 0;
+  int trunc_fname = 0;
+  int trunc_path = 0;
+  int hard_links = 0;
+  bool print_attributes = false;
+} testfind_stats;
+
+
+static int handleFile(JobControlRecord* jcr,
+                      FindFilesPacket* ff_pkt,
+                      bool top_level)
+{
+  if (!SaveFile(jcr, ff_pkt, top_level)) { return 0; }
+  if (!PrintFile(jcr, ff_pkt, top_level)) { return 0; }
+  return 1;
+}
+
+void ProcessFileset(directordaemon::FilesetResource* director_fileset,
+                    const char* configfile,
+                    bool print_attrs)
+{
+  testfind_stats.print_attributes = print_attrs;
+
+  my_config = InitFdConfig(configfile, M_ERROR_TERM);
+  my_config->ParseConfig();
+
+  me = static_cast<ClientResource*>(my_config->GetNextRes(R_CLIENT, nullptr));
+  no_signals = true;
+  me->compatible = true;
+
+  if (!CheckResources()) {
+    printf("Problem checking resources!");
+    return;
+  }
+
+  JobControlRecord* jcr;
+  EmptySocket* dird_sock = new EmptySocket;
+  jcr = create_new_director_session(dird_sock);
+
+  EmptySocket* stored_sock = new EmptySocket;
+  jcr->store_bsock = stored_sock;
+  stored_sock->message_length = 0;
+
+  DummyFdFilesetSocket* filed_sock = new DummyFdFilesetSocket;
+
+  filed_sock->jcr = jcr;
+  jcr->file_bsock = filed_sock;
+
+  crypto_cipher_t cipher = CRYPTO_CIPHER_NONE;
+  GetWantedCryptoCipher(jcr, &cipher);
+
+  InitFileset(jcr);
+  directordaemon::SendIncludeExcludeItems(jcr, director_fileset);
+  TermFileset(jcr);
+
+  BlastDataToStorageDaemon(jcr, cipher, DEFAULT_NETWORK_BUFFER_SIZE,
+                           handleFile);
+
+  testfind_stats.hard_links = jcr->fd_impl->ff->linkhash->size();
+  printf(_("\n"
+           "Total files    : %d\n"
+           "Max file length: %d\n"
+           "Max path length: %d\n"
+           "Files truncated: %d\n"
+           "Paths truncated: %d\n"
+           "Hard links     : %d\n"),
+         testfind_stats.num_files, testfind_stats.max_file_len,
+         testfind_stats.max_path_len, testfind_stats.trunc_fname,
+         testfind_stats.trunc_path, testfind_stats.hard_links);
+
+  jcr->file_bsock->close();
+  delete jcr->file_bsock;
+  jcr->file_bsock = nullptr;
+  CleanupFileset(jcr);
+  FreeJcr(jcr);
+
+  TermMsg();
```

Comment:
`TermMsg()` seems to free some global resources.  Is it intentional that both `ProcessFileset` and the testfind main function call `TermMsg()` ?

Reply:
leftover from previous iteration.